qiskit/synthesis/discrete_basis/gate_sequence.py

```diff
@@ -0,0 +1,386 @@
+        """Remove a pair of indices that cancel each other and *do not* change the matrices."""
+        for index in list(indices[::-1]):
+            self.gates.pop(index)
+            self.labels.pop(index)
+
+        # restore name
+        self.name = " ".join(self.labels)
+
+    def __eq__(self, other: "GateSequence") -> bool:
+        """Check if this GateSequence is the same as the other GateSequence.
+
+        Args:
+            other: The GateSequence that will be compared to ``self``.
+
+        Returns:
+            True if ``other`` is equivalent to ``self``, false otherwise.
+
+        """
+        if not len(self.gates) == len(other.gates):
+            return False
+
+        for gate1, gate2 in zip(self.gates, other.gates):
+            if gate1 != gate2:
+                return False
+
+        if self.global_phase != other.global_phase:
+            return False
+
+        return True
+
+    def to_circuit(self):
+        """Convert to a circuit.
+
+        If no gates set but the product is not the identity, returns a circuit with a
+        unitary operation to implement the matrix.
+        """
+        if len(self.gates) == 0 and not np.allclose(self.product, np.identity(3)):
+            circuit = QuantumCircuit(1, global_phase=self.global_phase)
+            su2 = _convert_so3_to_su2(self.product)
+            circuit.unitary(su2, [0])
+            return circuit
+
+        circuit = QuantumCircuit(1, global_phase=self.global_phase)
+        for gate in self.gates:
+            circuit.append(gate, [0])
+
+        return circuit
+
+    def append(self, gate: Gate) -> "GateSequence":
+        """Append gate to the sequence of gates.
+
+        Args:
+            gate: The gate to be appended.
+
+        Returns:
+            GateSequence with ``gate`` appended.
+        """
+        # invalidate euler angles and name
+        self._eulers = None
+
+        # TODO: this recomputes the product whenever we append something, which could be more
+        # efficient by storing the current matrix and just multiplying the input gate to it
+        # self.product = convert_su2_to_so3(self._compute_product(self.gates))
+        matrix = np.array(gate, dtype=np.complex128)
+        su2, phase = _convert_u2_to_su2(matrix)
+        so3 = _convert_su2_to_so3(su2)
+
+        self.product = so3.dot(self.product)
+        self.product_su2 = su2.dot(self.product_su2)
+        self.global_phase = self.global_phase + phase
+
+        self.gates.append(gate)
+        if len(self.labels) > 0:
+            self.name += f" {gate.name}"
+        else:
+            self.name = gate.name
+        self.labels.append(gate.name)
+
+        self.matrices.append(matrix)
+
+        return self
+
+    def adjoint(self) -> "GateSequence":
+        """Get the complex conjugate."""
+        adjoint = GateSequence()
+        adjoint.gates = [gate.inverse() for gate in reversed(self.gates)]
```

Comment:
Is there a reason to just do `GateSequence([gate.inverse() for gate in reversed(self.gates)])`? I think this question holds for a bunch of places we're constructing new `GateSequence` objects.